core/src/main/java/org/elasticsearch/script/ScriptEngineRegistry.java

```diff
@@ -0,0 +1,91 @@
+        Objects.requireNonNull(registrations);
+        Map<Class<? extends ScriptEngineService>, List<String>> registeredScriptEngineServices = new HashMap<>();
+        Map<String, Class<? extends ScriptEngineService>> registeredLanguages = new HashMap<>();
+        for (ScriptEngineRegistration registration : registrations) {
+            List<String> languages =
+                registeredScriptEngineServices.putIfAbsent(registration.getScriptEngineService(), Collections.unmodifiableList(registration.getScriptEngineLanguages()));
+            if (languages != null) {
+                throw new IllegalArgumentException("script engine service [" + registration.getScriptEngineService() + "] already registered for languages [" + String.join(",", languages) + "]");
+            }
+
+            for (String language : registration.getScriptEngineLanguages()) {
+                Class<? extends ScriptEngineService> scriptEngineServiceClazz =
+                    registeredLanguages.putIfAbsent(language, registration.getScriptEngineService());
+                if (scriptEngineServiceClazz != null) {
+                    throw new IllegalArgumentException("scripting language [" + language + "] already registered for script engine service [" + scriptEngineServiceClazz.getCanonicalName() + "]");
+                }
+            }
+        }
+
+        this.registeredScriptEngineServices = Collections.unmodifiableMap(registeredScriptEngineServices);
+        this.registeredLanguages = Collections.unmodifiableMap(registeredLanguages);
+    }
+
+    public Iterable<Class<? extends ScriptEngineService>> getRegisteredScriptEngineServices() {
+        return registeredScriptEngineServices.keySet();
+    }
+
+    public List<String> getLanguages(Class<? extends ScriptEngineService> scriptEngineService) {
+        Objects.requireNonNull(scriptEngineService);
+        return registeredScriptEngineServices.get(scriptEngineService);
+    }
+
+    public Map<String, Class<? extends ScriptEngineService>> getRegisteredLanguages() {
+        return registeredLanguages;
+    }
+
+    public static class ScriptEngineRegistration {
+        private final Class<? extends ScriptEngineService> scriptEngineService;
+        private final List<String> scriptEngineLanguages;
+
+        public ScriptEngineRegistration(Class<? extends ScriptEngineService> scriptEngineService, List<String> scriptEngineLanguages) {
+            Objects.requireNonNull(scriptEngineService);
+            Objects.requireNonNull(scriptEngineLanguages);
+            if (scriptEngineLanguages.isEmpty()) {
+                throw new IllegalArgumentException("languages for script engine service [" + scriptEngineService.getCanonicalName() + "] should be non-empty");
+            }
+            this.scriptEngineService = scriptEngineService;
+            this.scriptEngineLanguages = scriptEngineLanguages;
+        }
+
+        public Class<? extends ScriptEngineService> getScriptEngineService() {
```

Comment:
I've reduced the visibility of this method to package private, so I don't think that the work to do this is necessary. This is really just internal machinery so that all the dynamic settings (nb. not dynamically-updatable settings) can be registered with the settings module.